File: services/api/internal/store/store.go
package store

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

var ErrNotFound = errors.New("not found")

type JobSource struct {
	Source    string `json:"source"`
	SourceURL string `json:"source_url"`
}

// Score mirrors one `scores` row (Week 6). Written by the agent's scoring
// pass (services/agent/jobsonar_agent/run.py:score_jobs); the API only
// reads it -- see docs/TRD.md §4.1's note on where normalisation/scoring
// reasoning lives. Nil on a Job means the agent hasn't scored it yet, not
// that it scored zero.
type Score struct {
	Composite     float64  `json:"composite"`
	SkillCov      float64  `json:"skill_cov"`
	Semantic      *float64 `json:"semantic,omitempty"`
	SeniorityFit  float64  `json:"seniority_fit"`
	LocationFit   float64  `json:"location_fit"`
	Recency       float64  `json:"recency"`
	Band          string   `json:"band"`
	MatchedSkills []string `json:"matched_skills"`
	MissingSkills []string `json:"missing_skills"`
}

type Job struct {
	ID            uuid.UUID      `json:"id"`
	Title         string         `json:"title"`
	Company       string         `json:"company"`
	Location      string         `json:"location"`
	Source        string         `json:"source"`
	SourceURL     string         `json:"source_url"`
	DescriptionMD string         `json:"description_md,omitempty"`
	SalaryMin     *float64       `json:"salary_min,omitempty"`
	SalaryMax     *float64       `json:"salary_max,omitempty"`
	Currency      string         `json:"currency,omitempty"`
	PostedAt      *time.Time     `json:"posted_at,omitempty"`
	Status        string         `json:"status"`
	LastSeenAt    time.Time      `json:"last_seen_at"`
	Sources       []JobSource    `json:"sources"`
	Application   *AppBrief      `json:"application,omitempty"`
	Score         *Score         `json:"score,omitempty"`
	HasAnalysis   bool           `json:"has_analysis"`
	Analysis      *Analysis      `json:"analysis,omitempty"`
	Review        *CompanyReview `json:"review,omitempty"`
}

// JobListOpts filters and reorders GET /jobs. Ranking stays in SQL.
type JobListOpts struct {
	HasSalary bool   // only rows with salary_min or salary_max
	Sort      string // "" / "match" (composite) or "salary" (high pay first)
}

type Analysis struct {
	JustificationMD string    `json:"justification_md"`
	TailoringMD     string    `json:"tailoring_md"`
	Model           string    `json:"model"`
	CreatedAt       time.Time `json:"created_at"`
}

type AppBrief struct {
	ID     uuid.UUID `json:"id"`
	Status string    `json:"status"`
}

type Resume struct {
	ID         uuid.UUID `json:"id"`
	Status     string    `json:"status"`
	Error      string    `json:"error,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
	StorageURI string    `json:"-"`
}

type Profile struct {
	ID           uuid.UUID `json:"id"`
	Skills       []string  `json:"skills"`
	HasEmbedding bool      `json:"has_embedding"`
	Embedding    []float64 `json:"-"`
	LatestResume *Resume   `json:"latest_resume,omitempty"`
	UpdatedAt    time.Time `json:"updated_at"`
}

type Application struct {
	ID            uuid.UUID  `json:"id"`
	JobID         uuid.UUID  `json:"job_id"`
	Title         string     `json:"title"`
	Company       string     `json:"company"`
	Location      string     `json:"location"`
	SourceURL     string     `json:"source_url"`
	ResumeVariant string     `json:"resume_variant"`
	Status        string     `json:"status"`
	AppliedAt     *time.Time `json:"applied_at,omitempty"`
	Notes         string     `json:"notes"`
	CreatedAt     time.Time  `json:"created_at"`
}

type Company struct {
	ID         uuid.UUID `json:"id"`
	Name       string    `json:"name"`
	ATS        string    `json:"ats"`
	BoardToken string    `json:"board_token"`
	CreatedAt  time.Time `json:"created_at"`
}

type Store struct {
	pool *pgxpool.Pool
}

func New(ctx context.Context, dsn string) (*Store, error) {
	pool, err := pgxpool.New(ctx, dsn)
	if err != nil {
		return nil, fmt.Errorf("connect: %w", err)
	}
	if err := pool.Ping(ctx); err != nil {
		pool.Close()
		return nil, fmt.Errorf("ping: %w", err)
	}
	return &Store{pool: pool}, nil
}

func (s *Store) Close() { s.pool.Close() }

// currentProfileID names the same "single profile, most recently updated"
// selection used throughout (upsert_skills/upsert_profile in the agent;
// the one profile this single-user project targets, per CLAUDE.md).
const currentProfileID = `(SELECT id FROM profiles ORDER BY updated_at DESC LIMIT 1)`

func (s *Store) ListJobs(ctx context.Context, opts JobListOpts) ([]Job, error) {
	q := jobSelect + jobFrom + `
		WHERE sc.band IS DISTINCT FROM 'excluded'`
	if opts.HasSalary {
		q += `
		AND (j.salary_min IS NOT NULL OR j.salary_max IS NOT NULL)`
	}
	q += jobOrder(opts.Sort)
	rows, err := s.pool.Query(ctx, q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanJobs(rows)
}

func (s *Store) GetJob(ctx context.Context, id uuid.UUID) (Job, error) {
	// Unlike ListJobs, an excluded (hard-gated) score is still returned
	// here -- a job someone links to directly should explain why it was
	// excluded, not disappear (Week 6 Day 4: never silently dropped).
	row := s.pool.QueryRow(ctx, jobSelect+jobFrom+`
		WHERE j.id = $1
	`, id)
	j, err := scanJob(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return Job{}, ErrNotFound
	}
	return j, err
}

const jobSelect = `
		SELECT j.id, j.title, j.company, j.location, j.source, j.source_url,
		       j.description_md, j.salary_min, j.salary_max, COALESCE(j.currency, ''),
		       j.posted_at, j.status, j.last_seen_at,
		       srcs.sources,
		       a.id, a.status,
		       sc.composite, sc.skill_cov, sc.semantic, sc.seniority_fit,
		       sc.location_fit, sc.recency, sc.band, sc.matched_skills, sc.missing_skills,
		       COALESCE(an.has_analysis, false),
		       an.justification_md, an.tailoring_md, an.model, an.created_at,
		       rev.rating, COALESCE(rev.review_count, 0), COALESCE(rev.summary, ''),
		       rev.snippets, rev.links, COALESCE(rev.provider, ''), COALESCE(rev.status, ''),
		       rev.fetched_at
`

const jobFrom = `
		FROM jobs j
		LEFT JOIN applications a ON a.job_id = j.id
		LEFT JOIN LATERAL (
			SELECT COALESCE(json_agg(json_build_object('source', src.source, 'source_url', src.source_url)), '[]') AS sources
			FROM job_sources src WHERE src.job_id = j.id
		) srcs ON true
		LEFT JOIN LATERAL (
			SELECT * FROM scores sc
			WHERE sc.job_id = j.id AND sc.profile_id = ` + currentProfileID + `
			LIMIT 1
		) sc ON true
		LEFT JOIN LATERAL (
			SELECT true AS has_analysis, justification_md, tailoring_md, model, created_at
			FROM analyses an
			WHERE an.job_id = j.id AND an.profile_id = ` + currentProfileID + `
			LIMIT 1
		) an ON true
		LEFT JOIN company_reviews rev
		  ON rev.company_key = lower(trim(j.company))
		 AND rev.role_key = lower(trim(j.title))
`

// salaryUSDExpr converts posted ranges to an approximate USD figure so
// INR / EUR / GBP ranks compare. Missing currency + an India location
// is treated as INR (Adzuna IN often omits currency). Static FX is for
// ORDER BY only — the API still returns the source numbers.
const salaryUSDExpr = `COALESCE(j.salary_max, j.salary_min)
		* CASE
			WHEN (UPPER(TRIM(COALESCE(j.currency, ''))) IN ('INR', 'RS')
			      OR (COALESCE(TRIM(j.currency), '') = ''
			          AND j.location ~* '(india|pune|bengaluru|bangalore|hyderabad|mumbai|chennai|noida|gurgaon|gurugram|delhi|kolkata)'))
			     AND COALESCE(j.salary_max, j.salary_min) < 1000
			THEN 100000
			ELSE 1
		  END
		* CASE
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) IN ('INR', 'RS') THEN 0.012
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) IN ('EUR') THEN 1.08
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) IN ('GBP') THEN 1.27
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) = 'CAD' THEN 0.73
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) = 'AUD' THEN 0.66
			WHEN UPPER(TRIM(COALESCE(j.currency, ''))) = 'SGD' THEN 0.74
			WHEN COALESCE(TRIM(j.currency), '') = ''
			     AND j.location ~* '(india|pune|bengaluru|bangalore|hyderabad|mumbai|chennai|noida|gurgaon|gurugram|delhi|kolkata)'
			     THEN 0.012
			ELSE 1.0
		  END`

func jobOrder(sort string) string {
	if strings.EqualFold(sort, "salary") {
		return `
		ORDER BY ` + salaryUSDExpr + ` DESC NULLS LAST,
		         rev.rating DESC NULLS LAST,
		         sc.composite DESC NULLS LAST,
		         j.last_seen_at DESC`
	}
	return `
		ORDER BY sc.composite DESC NULLS LAST, j.last_seen_at DESC`
}

func (s *Store) CreateCompany(ctx context.Context, name, ats, token string) (Company, error) {
	var c Company
	err := s.pool.QueryRow(ctx, `
		INSERT INTO companies (name, ats, board_token)
		VALUES ($1, $2, $3)
		ON CONFLICT (ats, board_token) DO UPDATE SET name = EXCLUDED.name
		RETURNING id, name, ats, board_token, created_at
	`, name, ats, token).Scan(&c.ID, &c.Name, &c.ATS, &c.BoardToken, &c.CreatedAt)
	return c, err
}

type rowScanner interface {
	Scan(dest ...any) error
}

func scanJob(row rowScanner) (Job, error) {
	var j Job
	var posted *time.Time
	var sources []byte
	var appID *uuid.UUID
	var appStatus *string
	var composite, skillCov, seniorityFit, locationFit, recency *float64
	var semantic *float64
	var band *string
	var matchedRaw, missingRaw []byte
	var hasAnalysis bool
	var just, tail, model *string
	var analysisAt *time.Time
	var revRating *float64
	var revCount int
	var revSummary, revProvider, revStatus string
	var revSnips, revLinks []byte
	var revFetched *time.Time
	if err := row.Scan(&j.ID, &j.Title, &j.Company, &j.Location, &j.Source, &j.SourceURL,
		&j.DescriptionMD, &j.SalaryMin, &j.SalaryMax, &j.Currency, &posted, &j.Status, &j.LastSeenAt, &sources, &appID, &appStatus,
		&composite, &skillCov, &semantic, &seniorityFit, &locationFit, &recency, &band, &matchedRaw, &missingRaw,
		&hasAnalysis, &just, &tail, &model, &analysisAt,
		&revRating, &revCount, &revSummary, &revSnips, &revLinks, &revProvider, &revStatus, &revFetched); err != nil {
		return Job{}, err
	}
	j.HasAnalysis = hasAnalysis
	j.PostedAt = posted
	if err := jsonUnmarshalSources(sources, &j.Sources); err != nil {
		return Job{}, err
	}
	if j.Sources == nil {
		j.Sources = []JobSource{}
	}
	if appID != nil && appStatus != nil {
		j.Application = &AppBrief{ID: *appID, Status: *appStatus}
	}
	if composite != nil && band != nil {
		sc := &Score{
			Composite: *composite, Semantic: semantic, Band: *band,
		}
		if skillCov != nil {
			sc.SkillCov = *skillCov
		}
		if seniorityFit != nil {
			sc.SeniorityFit = *seniorityFit
		}
		if locationFit != nil {
			sc.LocationFit = *locationFit
		}
		if recency != nil {
			sc.Recency = *recency
		}
		if err := unmarshalJSON(matchedRaw, &sc.MatchedSkills); err != nil {
			return Job{}, err
		}
		if err := unmarshalJSON(missingRaw, &sc.MissingSkills); err != nil {
			return Job{}, err
		}
		if sc.MatchedSkills == nil {
			sc.MatchedSkills = []string{}
		}
		if sc.MissingSkills == nil {
			sc.MissingSkills = []string{}
		}
		j.Score = sc
	}
	if just != nil || tail != nil {
		a := &Analysis{}
		if just != nil {
			a.JustificationMD = *just
		}
		if tail != nil {
			a.TailoringMD = *tail
		}
		if model != nil {
			a.Model = *model
		}
		if analysisAt != nil {
			a.CreatedAt = *analysisAt
		}
		j.Analysis = a
		j.HasAnalysis = true
	}
	if revStatus != "" || len(revLinks) > 0 {
		rev := &CompanyReview{
			Company:     j.Company,
			RoleTitle:   j.Title,
			Rating:      revRating,
			ReviewCount: revCount,
			Summary:     revSummary,
			Provider:    revProvider,
			Status:      revStatus,
		}
		if err := unmarshalJSON(revSnips, &rev.Snippets); err != nil && len(revSnips) > 0 {
			return Job{}, err
		}
		if rev.Snippets == nil {
			rev.Snippets = []ReviewSnippet{}
		}
		if len(revLinks) > 0 {
			if err := unmarshalJSON(revLinks, &rev.Links); err != nil {
				return Job{}, err
			}
		}
		if revFetched != nil {
			rev.FetchedAt = *revFetched
		}
		j.Review = rev
	}
	return j, nil
}

func scanJobs(rows pgx.Rows) ([]Job, error) {
	var out []Job
	for rows.Next() {
		j, err := scanJob(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, j)
	}
	if out == nil {
		out = []Job{}
	}
	return out, rows.Err()
}

func jsonUnmarshalSources(b []byte, dest *[]JobSource) error {
	if len(b) == 0 {
		*dest = []JobSource{}
		return nil
	}
	return unmarshalJSON(b, dest)
}

func FormatVector(v []float64) string {
	var b strings.Builder
	b.WriteByte('[')
	for i, x := range v {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteString(strconv.FormatFloat(x, 'f', -1, 64))
	}
	b.WriteByte(']')
	return b.String()
}

func ParseVector(s string) ([]float64, error) {
	s = strings.TrimSpace(s)
	s = strings.TrimPrefix(s, "[")
	s = strings.TrimSuffix(s, "]")
	if s == "" {
		return nil, nil
	}
	parts := strings.Split(s, ",")
	out := make([]float64, 0, len(parts))
	for _, p := range parts {
		x, err := strconv.ParseFloat(strings.TrimSpace(p), 64)
		if err != nil {
			return nil, err
		}
		out = append(out, x)
	}
	return out, nil
}

File: services/agent/jobsonar_agent/store.py
from __future__ import annotations

import json
from collections.abc import Iterable

import psycopg

from jobsonar_agent import config


def _vec(values: Iterable[float]) -> str:
    return "[" + ",".join(f"{x:.8f}" for x in values) + "]"


def _json_list(value) -> list:
    if value is None:
        return []
    if isinstance(value, str):
        value = json.loads(value)
    return list(value) if value else []


# Hard gates live in this CASE (CLAUDE.md golden rule 3 / FRD FR-12):
# must-have skills, seniority mismatch, or location mismatch force
# band='excluded' regardless of the composite the agent computed.
_SCORE_INSERT = """
INSERT INTO scores (job_id, profile_id, composite, skill_cov, semantic,
                     seniority_fit, location_fit, recency, band,
                     matched_skills, missing_skills, scored_at)
SELECT
    %(job_id)s::uuid, %(profile_id)s::uuid, %(composite)s, %(skill_cov)s, %(semantic)s,
    %(seniority_fit)s, %(location_fit)s, %(recency)s,
    CASE
        WHEN NOT (p.must_have_skills = '[]'::jsonb
                  OR p.must_have_skills <@ COALESCE(j.skills_extracted, '[]'::jsonb))
            THEN 'excluded'
        WHEN COALESCE(p.seniority, '') <> '' AND %(seniority_fit)s < 0.3
            THEN 'excluded'
        WHEN (COALESCE(p.location, '') <> '' OR COALESCE(p.remote_pref, '') <> '')
             AND %(location_fit)s = 0.0
            THEN 'excluded'
        ELSE %(band)s
    END,
    %(matched_skills)s::jsonb, %(missing_skills)s::jsonb, now()
FROM jobs j, profiles p
WHERE j.id = %(job_id)s::uuid AND p.id = %(profile_id)s::uuid
ON CONFLICT (job_id, profile_id) DO UPDATE SET
    composite = EXCLUDED.composite,
    skill_cov = EXCLUDED.skill_cov,
    semantic = EXCLUDED.semantic,
    seniority_fit = EXCLUDED.seniority_fit,
    location_fit = EXCLUDED.location_fit,
    recency = EXCLUDED.recency,
    band = EXCLUDED.band,
    matched_skills = EXCLUDED.matched_skills,
    missing_skills = EXCLUDED.missing_skills,
    scored_at = EXCLUDED.scored_at
"""


class Store:
    def __init__(self, dsn: str | None = None):
        self.dsn = dsn or config.POSTGRES_DSN

    def connect(self):
        return psycopg.connect(self.dsn)

    def pending_resumes(self) -> list[dict]:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT id::text, storage_uri
                FROM resumes
                WHERE status = 'pending'
                ORDER BY created_at
                """
            )
            return [{"id": r[0], "storage_uri": r[1]} for r in cur.fetchall()]

    def mark_resume(self, resume_id: str, status: str, parsed: dict | None, error: str = "") -> None:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                UPDATE resumes
                SET status = %s, parsed = %s, error = %s
                WHERE id = %s::uuid
                """,
                (status, json.dumps(parsed) if parsed is not None else None, error, resume_id),
            )
            conn.commit()

    def upsert_skills(self, skills: list[str]) -> None:
        raw = json.dumps(skills)
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute("SELECT id FROM profiles ORDER BY updated_at DESC LIMIT 1")
            row = cur.fetchone()
            if row:
                cur.execute(
                    """
                    UPDATE profiles
                    SET skills = %s::jsonb, embedding = NULL, updated_at = now()
                    WHERE id = %s
                    """,
                    (raw, row[0]),
                )
            else:
                cur.execute("INSERT INTO profiles (skills) VALUES (%s::jsonb)", (raw,))
            conn.commit()

    def upsert_profile(self, skills: list[str], embedding: list[float]) -> None:
        raw = json.dumps(skills)
        vec = _vec(embedding)
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute("SELECT id FROM profiles ORDER BY updated_at DESC LIMIT 1")
            row = cur.fetchone()
            if row:
                cur.execute(
                    """
                    UPDATE profiles
                    SET skills = %s::jsonb, embedding = %s::vector, updated_at = now()
                    WHERE id = %s
                    """,
                    (raw, vec, row[0]),
                )
            else:
                cur.execute(
                    """
                    INSERT INTO profiles (skills, embedding)
                    VALUES (%s::jsonb, %s::vector)
                    """,
                    (raw, vec),
                )
            conn.commit()

    def profiles_missing_embedding(self) -> list[dict]:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT id::text, skills
                FROM profiles
                WHERE embedding IS NULL
                """
            )
            out = []
            for rid, skills in cur.fetchall():
                if isinstance(skills, str):
                    skills = json.loads(skills)
                out.append({"id": rid, "skills": skills or []})
            return out

    def set_profile_embedding(self, profile_id: str, embedding: list[float]) -> None:
        # Embed-only: do not bump updated_at. That column stale-marks every
        # scores row; a vector fill-in must not force a full rescore.
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                UPDATE profiles
                SET embedding = %s::vector
                WHERE id = %s::uuid
                """,
                (_vec(embedding), profile_id),
            )
            conn.commit()

    def jobs_missing_embeddings(self, limit: int) -> list[dict]:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT j.id::text, j.title, j.description_md
                FROM jobs j
                LEFT JOIN job_embeddings e ON e.job_id = j.id
                WHERE e.job_id IS NULL
                ORDER BY j.last_seen_at DESC
                LIMIT %s
                """,
                (limit,),
            )
            return [
                {"id": r[0], "title": r[1] or "", "description_md": r[2] or ""}
                for r in cur.fetchall()
            ]

    def current_profile(self) -> dict | None:
        """The single profile this project targets (CLAUDE.md: single-user).
        Same "most recently updated" selection as upsert_skills/upsert_profile."""
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT id::text, skills, seniority, location, remote_pref
                FROM profiles
                ORDER BY updated_at DESC
                LIMIT 1
                """
            )
            row = cur.fetchone()
            if not row:
                return None
            rid, skills, seniority, location, remote_pref = row
            if isinstance(skills, str):
                skills = json.loads(skills)
            return {
                "id": rid,
                "skills": skills or [],
                "seniority": seniority,
                "location": location,
                "remote_pref": remote_pref,
            }

    def jobs_for_scoring(self, profile_id: str, limit: int) -> list[dict]:
        """Jobs missing a scores row, or re-scored since last_seen_at/
        profile.updated_at moved -- same "backfill what's stale" shape as
        jobs_missing_embeddings."""
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT j.id::text, j.title, j.description_md, j.location, j.remote_type,
                       j.posted_at, j.first_seen_at, j.skills_extracted,
                       (1 - (e.embedding <=> p.embedding))::float8 AS semantic
                FROM jobs j
                LEFT JOIN job_embeddings e ON e.job_id = j.id
                JOIN profiles p ON p.id = %(profile_id)s::uuid
                LEFT JOIN scores s ON s.job_id = j.id AND s.profile_id = p.id
                WHERE s.job_id IS NULL
                   OR s.scored_at < j.last_seen_at
                   OR s.scored_at < p.updated_at
                ORDER BY j.last_seen_at DESC
                LIMIT %(limit)s
                """,
                {"profile_id": profile_id, "limit": limit},
            )
            cols = (
                "id", "title", "description_md", "location", "remote_type",
                "posted_at", "first_seen_at", "skills_extracted", "semantic",
            )
            out = []
            for row in cur.fetchall():
                item = dict(zip(cols, row))
                item["skills_extracted"] = (
                    None if item["skills_extracted"] is None
                    else _json_list(item["skills_extracted"])
                )
                out.append(item)
            return out

    def upsert_score(
        self,
        job_id: str,
        profile_id: str,
        *,
        composite: float,
        skill_cov: float,
        semantic: float | None,
        seniority_fit: float,
        location_fit: float,
        recency: float,
        band: str,
        matched_skills: list[str],
        missing_skills: list[str],
    ) -> None:
        """Persists one job's sub-scores. Hard gates are the CASE in
        _SCORE_INSERT — decided by Postgres, not Python control flow.
        """
        self.write_score_batch(profile_id, [{
            "job_id": job_id,
            "composite": composite,
            "skill_cov": skill_cov,
            "semantic": semantic,
            "seniority_fit": seniority_fit,
            "location_fit": location_fit,
            "recency": recency,
            "band": band,
            "matched_skills": matched_skills,
            "missing_skills": missing_skills,
        }])

    def write_score_batch(self, profile_id: str, rows: list[dict]) -> None:
        """One connection / one commit for a scoring batch. Optional
        write_skills updates jobs.skills_extracted before the gate CASE
        reads it.
        """
        if not rows:
            return
        with self.connect() as conn, conn.cursor() as cur:
            for row in rows:
                if row.get("write_skills"):
                    cur.execute(
                        "UPDATE jobs SET skills_extracted = %s::jsonb WHERE id = %s::uuid",
                        (json.dumps(row.get("skills") or []), row["job_id"]),
                    )
                cur.execute(
                    _SCORE_INSERT,
                    {
                        "job_id": row["job_id"],
                        "profile_id": profile_id,
                        "composite": row["composite"],
                        "skill_cov": row["skill_cov"],
                        "semantic": row.get("semantic"),
                        "seniority_fit": row["seniority_fit"],
                        "location_fit": row["location_fit"],
                        "recency": row["recency"],
                        "band": row["band"],
                        "matched_skills": json.dumps(row.get("matched_skills") or []),
                        "missing_skills": json.dumps(row.get("missing_skills") or []),
                    },
                )
            conn.commit()

    def set_job_skills_extracted(self, job_id: str, skills: list[str]) -> None:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                "UPDATE jobs SET skills_extracted = %s::jsonb WHERE id = %s::uuid",
                (json.dumps(skills), job_id),
            )
            conn.commit()

    def jobs_for_deep_dive(self, profile_id: str, band: str) -> list[dict]:
        """Shortlist: scores at `band` with no analyses row yet.
        A rescore must not rewrite existing prose (that made every
        resume upload re-prompt every strong job)."""
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                SELECT j.id::text, j.title, j.company, j.description_md,
                       s.matched_skills, s.missing_skills, s.band
                FROM scores s
                JOIN jobs j ON j.id = s.job_id
                LEFT JOIN analyses a ON a.job_id = s.job_id AND a.profile_id = s.profile_id
                WHERE s.profile_id = %s::uuid
                  AND s.band = %s
                  AND a.job_id IS NULL
                ORDER BY s.composite DESC
                """,
                (profile_id, band),
            )
            out = []
            for rid, title, company, desc, matched, missing, jband in cur.fetchall():
                if isinstance(matched, str):
                    matched = json.loads(matched)
                if isinstance(missing, str):
                    missing = json.loads(missing)
                out.append({
                    "id": rid,
                    "title": title or "",
                    "company": company or "",
                    "description_md": desc or "",
                    "matched_skills": matched or [],
                    "missing_skills": missing or [],
                    "band": jband,
                })
            return out

    def upsert_analysis(
        self,
        job_id: str,
        profile_id: str,
        *,
        justification_md: str,
        tailoring_md: str,
        model: str,
    ) -> None:
        with self.connect() as conn, conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO analyses (job_id, profile_id, justification_md, tailoring_md, model, created_at)
                VALUES (%s::uuid, %s::uuid, %s, %s, %s, now())
                ON CONFLICT (job_id, profile_id) DO UPDATE SET
                    justification_md = EXCLUDED.justification_md,
                    tailoring_md = EXCLUDED.tailoring_md,
                    model = EXCLUDED.model,
                    created_at = EXCLUDED.created_at
                """,
                (job_id, profile_id, justification_md, tailoring_md, model),
            )
            conn.commit()

    def upsert_job_embeddings(self, rows: list[tuple[str, list[float]]], model: str) -> None:
        if not rows:
            return
        with self.connect() as conn, conn.cursor() as cur:
            for job_id, embedding in rows:
                cur.execute(
                    """
                    INSERT INTO job_embeddings (job_id, embedding, model)
                    VALUES (%s::uuid, %s::vector, %s)
                    ON CONFLICT (job_id) DO UPDATE
                    SET embedding = EXCLUDED.embedding,
                        model = EXCLUDED.model,
                        updated_at = now()
                    """,
                    (job_id, _vec(embedding), model),
                )
            conn.commit()
